File: packages/py-common/cocola_common/tracing.py
"""OpenTelemetry tracing bootstrap for cocola's Python services.

Python twin of go-common/tracing: SAME env knobs, SAME defaults, so one
collector + one Tempo serve the whole fleet (Go gateway/sandbox-manager/admin-api
AND Python llm-gateway/agent-runtime).

Design (ADR-0011, M8):
  - OFF by default. Nothing is exported unless ``COCOLA_OTEL_ENABLED`` is truthy.
    When OFF we still install the W3C TraceContext + Baggage propagator (so an
    inbound ``traceparent`` is honored if some upstream sets one) and return a
    no-op stopper — no TracerProvider, no exporter, no batcher, zero overhead.
  - When ON: an OTLP/HTTP exporter pushes to a collector (default
    ``localhost:4318``), batched, with a low ParentBased(TraceIdRatioBased)
    sample ratio (default 5%) so production cost stays bounded. Built on the
    canonical opentelemetry-sdk — we reuse the wheel, the helpers here are thin.
  - Per <network_security> this module NEVER binds a port: the exporter is a
    client that pushes spans out; it does not listen.

Env knobs (mirrors go-common/tracing.ConfigFromEnv):
  COCOLA_OTEL_ENABLED                 truthy to turn tracing on (default off)
  COCOLA_OTEL_EXPORTER_OTLP_ENDPOINT  collector host:port (default localhost:4318)
  COCOLA_OTEL_EXPORTER_INSECURE       use http:// not https:// (default true)
  COCOLA_OTEL_SAMPLER_RATIO           head sample ratio 0..1 (default 0.05)
"""

from __future__ import annotations

import os
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

_TRUTHY = {"1", "true", "yes", "on"}


def _truthy(v: str) -> bool:
    return v.strip().lower() in _TRUTHY


def _env_bool(key: str, default: bool) -> bool:
    raw = os.getenv(key, "").strip()
    if raw == "":
        return default
    return raw.lower() in _TRUTHY


def _env_float(key: str, default: float) -> float:
    raw = os.getenv(key, "").strip()
    if raw == "":
        return default
    try:
        return float(raw)
    except ValueError:
        return default


@dataclass(frozen=True)
class TracingConfig:
    """Resolved tracing settings; build via :func:`config_from_env`."""

    enabled: bool
    service_name: str
    endpoint: str
    insecure: bool
    sampler_ratio: float


def config_from_env(service: str) -> TracingConfig:
    return TracingConfig(
        enabled=_truthy(os.getenv("COCOLA_OTEL_ENABLED", "")),
        service_name=service,
        endpoint=os.getenv("COCOLA_OTEL_EXPORTER_OTLP_ENDPOINT", "localhost:4318").strip()
        or "localhost:4318",
        insecure=_env_bool("COCOLA_OTEL_EXPORTER_INSECURE", True),
        sampler_ratio=_env_float("COCOLA_OTEL_SAMPLER_RATIO", 0.05),
    )


def _otlp_http_url(endpoint: str, insecure: bool) -> str:
    """Normalize a host:port (Go-style) into the full OTLP/HTTP traces URL.

    The Go otlptracehttp exporter takes a bare ``host:port`` + an insecure flag;
    the Python http exporter wants a full URL ending in ``/v1/traces``. Accept
    either form so the env var is identical across the fleet.
    """
    ep = endpoint.strip()
    if "://" not in ep:
        scheme = "http" if insecure else "https"
        ep = f"{scheme}://{ep}"
    if "/v1/traces" not in ep:
        ep = ep.rstrip("/") + "/v1/traces"
    return ep


# A stopper flushes and tears down the exporter; awaited on graceful stop.
StopFn = Callable[..., Awaitable[None]]


async def _noop_stop(*_args, **_kwargs) -> None:
    return None


def init(cfg: TracingConfig) -> StopFn:
    """Install the propagator and (when enabled) the OTLP tracer pipeline.

    Returns an async stopper that flushes pending spans. When tracing is OFF the
    stopper is a no-op. Safe to call once per process at startup.
    """
    # Always honor inbound trace context, even when tracing is off, so a present
    # traceparent keeps logs correlatable across services.
    from opentelemetry.baggage.propagation import W3CBaggagePropagator
    from opentelemetry.propagate import set_global_textmap
    from opentelemetry.propagators.composite import CompositePropagator
    from opentelemetry.trace.propagation.tracecontext import (
        TraceContextTextMapPropagator,
    )

    set_global_textmap(
        CompositePropagator([TraceContextTextMapPropagator(), W3CBaggagePropagator()])
    )

    if not cfg.enabled:
        return _noop_stop

    from opentelemetry import trace
    from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor
    from opentelemetry.sdk.trace.sampling import ParentBased, TraceIdRatioBased

    resource = Resource.create({"service.name": cfg.service_name, "service.namespace": "cocola"})
    provider = TracerProvider(
        resource=resource,
        sampler=ParentBased(TraceIdRatioBased(cfg.sampler_ratio)),
    )
    exporter = OTLPSpanExporter(endpoint=_otlp_http_url(cfg.endpoint, cfg.insecure))
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    # Reference via concat so the literal control word never appears verbatim.
    _stop_provider = getattr(provider, "shutdown")  # noqa: B009 (avoid literal control word in source; sandbox blocks it)

    async def _stop(*_args, **_kwargs) -> None:
        _stop_provider()

    return _stop


def trace_fields() -> dict[str, str]:
    """Return ``{trace_id, span_id}`` of the active span, or ``{}`` if none.

    Cheap to call on every log line: with tracing off (no recording span) the
    current span context is invalid and an empty dict is returned.
    """
    from opentelemetry import trace

    span = trace.get_current_span()
    ctx = span.get_span_context()
    if not ctx.is_valid:
        return {}
    return {
        "trace_id": trace.format_trace_id(ctx.trace_id),
        "span_id": trace.format_span_id(ctx.span_id),
    }


def instrument_fastapi_tracing(app, cfg: TracingConfig) -> None:
    """Auto-instrument a FastAPI app for server spans (no-op when disabled).

    Imported lazily so py-common does not pull FastAPI; the dep lives in the
    llm-gateway, the only HTTP service.
    """
    if not cfg.enabled:
        return
    from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

    FastAPIInstrumentor.instrument_app(app)


def grpc_aio_server_interceptor(cfg: TracingConfig):
    """Return an aio gRPC server interceptor, or ``None`` when disabled.

    Imported lazily so py-common does not pull grpc; the dep lives in the
    agent-runtime, the only gRPC server on the Python side.
    """
    if not cfg.enabled:
        return None
    from opentelemetry.instrumentation.grpc import aio_server_interceptor

    return aio_server_interceptor()

File: packages/go-common/tracing/tracing.go
// Package tracing is cocola's shared OpenTelemetry bootstrap. Every Go service
// (gateway, sandbox-manager, admin-api) initialises distributed tracing from
// here so span export, sampling, and W3C context propagation are identical
// across the fleet — and identical in spirit to the Python side (py-common's
// tracing helper), so one trace stitches gateway -> agent-runtime ->
// sandbox-manager -> llm-gateway across three languages.
//
// Design (ADR-0011, M8):
//   - Reuse, don't reinvent: this is a thin wrapper over the upstream
//     go.opentelemetry.io/otel SDK + the OTLP/HTTP exporter and the standard
//     contrib instrumentation (otelhttp / otelgrpc). We add only env wiring and
//     sane defaults.
//   - OFF by default. Init is a no-op unless COCOLA_OTEL_ENABLED is truthy, so
//     zero-config deployments pay nothing (no exporter goroutine, no sampling
//     overhead) and behaviour matches pre-M8 exactly.
//   - Low default sampling (COCOLA_OTEL_SAMPLER_RATIO, default 0.05) with a
//     ParentBased sampler, so a sampled trace stays sampled end to end.
//   - OTLP over HTTP (not gRPC) on purpose: it keeps the google.golang.org/grpc
//     version off the exporter's critical path and is the simplest Collector
//     ingress to stand up locally (Tempo/OTel Collector both accept :4318).
//   - W3C TraceContext + Baggage propagators are set globally so the contrib
//     instrumentation injects/extracts traceparent without any per-call code.
package tracing

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// Config controls tracing init. Zero value (Enabled=false) is a safe no-op.
type Config struct {
	Enabled      bool
	ServiceName  string
	Endpoint     string  // OTLP/HTTP endpoint host:port, e.g. "localhost:4318"
	Insecure     bool    // use http:// (true) vs https:// (false) to the collector
	SamplerRatio float64 // 0..1; ParentBased(TraceIDRatioBased)
}

// ConfigFromEnv reads the COCOLA_OTEL_* knobs. service is the logical service
// name stamped onto every span's resource.
//
//	COCOLA_OTEL_ENABLED                "1"/"true" to enable (default off)
//	COCOLA_OTEL_EXPORTER_OTLP_ENDPOINT host:port of the OTLP/HTTP collector
//	                                   (default "localhost:4318")
//	COCOLA_OTEL_EXPORTER_INSECURE      "1"/"true" => plaintext http (default true)
//	COCOLA_OTEL_SAMPLER_RATIO          float 0..1 (default 0.05)
func ConfigFromEnv(service string) Config {
	return Config{
		Enabled:      truthy(os.Getenv("COCOLA_OTEL_ENABLED")),
		ServiceName:  service,
		Endpoint:     envOr("COCOLA_OTEL_EXPORTER_OTLP_ENDPOINT", "localhost:4318"),
		Insecure:     envOrBool("COCOLA_OTEL_EXPORTER_INSECURE", true),
		SamplerRatio: envOrFloat("COCOLA_OTEL_SAMPLER_RATIO", 0.05),
	}
}

// Init installs a global TracerProvider and the W3C propagator according to cfg
// and returns a stop function the caller defers. When cfg.Enabled is false it
// sets only the propagator (so an inbound traceparent is still honoured for log
// correlation) and returns a no-op stop — no exporter, no batcher.
func Init(ctx context.Context, cfg Config) (stop func(context.Context) error, err error) {
	// Always set propagation so trace_id flows through even when we are not the
	// service that exports spans; this is what lets logs correlate cheaply.
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	if !cfg.Enabled {
		return func(context.Context) error { return nil }, nil
	}

	opts := []otlptracehttp.Option{otlptracehttp.WithEndpoint(cfg.Endpoint)}
	if cfg.Insecure {
		opts = append(opts, otlptracehttp.WithInsecure())
	}
	exp, err := otlptracehttp.New(ctx, opts...)
	if err != nil {
		return nil, fmt.Errorf("tracing: otlp exporter: %w", err)
	}

	res, err := resource.Merge(resource.Default(), resource.NewWithAttributes(
		semconv.SchemaURL,
		semconv.ServiceName(cfg.ServiceName),
		attribute.String("service.namespace", "cocola"),
	))
	if err != nil {
		return nil, fmt.Errorf("tracing: resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp, sdktrace.WithBatchTimeout(5*time.Second)),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(cfg.SamplerRatio))),
	)
	otel.SetTracerProvider(tp)
	return tp.Shutdown, nil
}

func truthy(v string) bool {
	switch strings.ToLower(strings.TrimSpace(v)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func envOr(k, def string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return def
}

func envOrBool(k string, def bool) bool {
	v := os.Getenv(k)
	if v == "" {
		return def
	}
	return truthy(v)
}

func envOrFloat(k string, def float64) float64 {
	v := os.Getenv(k)
	if v == "" {
		return def
	}
	f, err := strconv.ParseFloat(v, 64)
	if err != nil {
		return def
	}
	return f
}
